### DISCUSSION TITLE
dispatch a custom event to pass a array to another x-data for template re-draw

### DISCUSSION THREAD

**Discussion by ericxin1982** [alpinejs/alpine#4131] [1. Help] [answered]
Hi,

I have two x-data scope, and one has init() to fetch api to get a menu structure and would pass an array to an object variable in another x-data scope through dispatch, but that object is {} when init(), and that object is used by template to do x-for "<"div">" generation, and I want to know if the variable has new value assignment, the template would run again or it does not work that way.

Thanks
Eric Xin

**Comment by joshhanley** [comment]
@ericxin1982 Can you please use the help category instead of general? As the general category can't be marked as answered. Thanks!

I have moved this one to help.

**Comment by ekwoka** [comment]
can you show some code?

I really cna't visualize what the issue is.

If the thing will be an array? why is it `{}`?

Why are you using dispatch and not a shared parent or store?

What issue are you actually running into right now? What is happening that you don't expect? what do you expect that isn't happening?

Can you demo the issue on https://awesomealpine.com/play ?

**Reply by ericxin1982** [reply]
right, not {}, it is []. 
Not a shared parent since that div was added afterward, and I am concerned which has css rule hooks it, so I do not plan to move that div into the same parent x-data, that is what I thought in the beginning.

I fulfill the code first, and if it works I let you know, if it does not work, I zip the folder and upload here, you help me take a look at it.

**Reply by ekwoka** [reply]
Nobody is gonna download and run your zip.

That's a good rule to learn.

Make it easy to help you.

I gave you a sandbox link you can put code in in a controlled way to demo.

**Reply by ericxin1982** [reply]
No, I do not worry about the code, I just do not know how to use the playground.
[index.zip](https://github.com/alpinejs/alpine/files/14893096/index.zip)
If my code is like this, how to put code there?

**Reply by SimoTod** [reply]
You need to make an effort and try to figure out how to use the playground if you need that much help. Downloading a random zip created by a person you don't know is a security risk.

You just copy paste a simplified version of your code on the playground and share it.

You can also use codepen or whatever you are more comfortable with.

**Reply by ericxin1982** [reply]
Ok, let me understand how to use the online playground first and next time I can share the link directly if I ask for help.

**Comment by ericxin1982** [comment, marked as answer]
Ok, it works. But what I tries to ask is the event timing issue.
The logic is like this, I have one x-data named "sidebarA" for sidebar, and "sidebarA" has one array named "menu", and the menu array is like this:
![image](https://github.com/alpinejs/alpine/assets/9047436/6bfa17f7-1882-46ba-bcd2-62ca2ce6e959)
Let's assume that is assigned by ajax get, and I also want to have another array (named spadiv) for menu id collection, I want to pass another x-data named "spacontainer" for spa individual div. Like this:
![image](https://github.com/alpinejs/alpine/assets/9047436/4241d76a-63fa-4f04-9e30-6390fcce0977)
From sequence, sidebar happened first, and container happened afterwards, just now what I asked is sidebarA this x-data init completed, menu array ready, and spadiv array ready, and I pass to another x-data spacontainer, and it can take it before init takes place? I am not so sure this point only. 
I forgot alpine.store, I just thought dispatch to inject value across two different x-data scope only, but I am doubt if I use dispatch that way whether it can match my expectation, right now I am using $store, that works fine. Please advise if needed!

My purpose is I try to avoid another ajax get for spadiv id array in spadiv x-data that scope.

**Reply by ekwoka** [reply]
I really still don't understand what you're describing.

**Reply by SimoTod** [reply]
If you have multiple components using exactly the same array and they should always be in sync, you probably want to look into using a store (https://alpinejs.dev/globals/alpine-store) instead of trying to sync them using events

**Reply by ericxin1982** [reply]
![sdfad_1](https://github.com/alpinejs/alpine/assets/9047436/fe1565e3-8661-4d23-8974-4ce4742f0f60)
![sdfad_2](https://github.com/alpinejs/alpine/assets/9047436/b5042b9b-f428-4942-b30e-800556521fc8)
![sdfad_3](https://github.com/alpinejs/alpine/assets/9047436/b080d00b-b8d0-43a7-9652-57b0c7496cbf)

I want to pass a array to another another x-data, my original thought is using event, but as Simoto or Ekwoka mentioned, I can use $store.
 I was just not sure whether the template uses array but the array had not been passed to yet. So my concern is the timing across different x-data scope.
![image](https://github.com/alpinejs/alpine/assets/9047436/180eaf1d-7256-41cf-aacd-ac2579aa5b43)
![image](https://github.com/alpinejs/alpine/assets/9047436/6d062bc4-f0fd-48b2-8ae3-9fb50ea39c9f)

If I just treat them as sync, then I am ok to do that through $store always.
I am very new to alpinejs, but I love it very much, I try to learn every single piece.

Thanks so much for all your help!

**Reply by ekwoka** [reply]
Yes, you can use a store. Initialize the value as an empty array.

The x-for will render nothing.

Then your request that gives the data will fill it in, and the x-for will then loop over and render it.
